torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -217,6 +217,145 @@ class ContigIDs : public OptInDispatch {
+        consumer_tv->getRootDomain().end(),
+        [&](IterDomain* consumer_root) {
+          auto concrete_id =
+              gpu_lower->caIndexMap().getConcreteMappedID(consumer_root);
+          auto it = reference_concrete_map.find(concrete_id);
+          return it != reference_concrete_map.end() &&
+              it->second == reference_root_axis;
+        });
+    // When no corresponding ID of the consumer exists, the reference
+    // axis can be ignored
+    if (consumer_it == consumer_tv->getRootDomain().end()) {
+      continue;
+    }
+    auto consumer_root_axis = *consumer_it;
+    auto root_axis_info =
+        gpu_lower->haloInfo().getRootAxisInfo(consumer_root_axis);
+    if (root_axis_info.width() == 0) {
+      continue;
+    }
+    halo_info.setRootAxisInfo(reference_root_axis, root_axis_info);
+  }
+
+  halo_info.build(reference_domain);
+
+  return;
+}
+
+// Get a map of IterDomains to halo-extended extents of corresponding
+// reference IterDomains.
+//
+// ref_map: ref-to-consumer in consumer indexing; ref-to-producer in
+// producer indexing
+std::unordered_map<kir::IterDomain*, kir::Val*> getReferenceHaloExtentMap(
+    const ReferenceTensor& reference,
+    const TensorView* consumer_tv,
+    const std::unordered_map<IterDomain*, IterDomain*>& ref_map,
+    const std::unordered_map<kir::IterDomain*, kir::Val*>& extent_map) {
+  const auto gpu_lower = GpuLower::current();
+
+  // First, update HaloInfo with the reference tensor, which reflects
+  // the halo extents of the consumer tensor.
+  updateHaloInfoForReference(reference, consumer_tv);
+
+  const auto& halo_info = gpu_lower->haloInfo();
+
+  std::unordered_map<kir::IterDomain*, kir::Val*> reference_halo_extent_map;
+
+  // Propagate halo extents of the reference to the consumer or
+  // producer tensor
+  for (auto kv : ref_map) {
+    auto ref_id = gpu_lower->lowerValue(kv.first)->as<kir::IterDomain>();
+    auto producer_or_consumer_id =
+        gpu_lower->lowerValue(kv.second)->as<kir::IterDomain>();
+    auto extent = halo_info.getExtent(ref_id);
+    if (extent == nullptr) {
+      auto extent_it = extent_map.find(ref_id);
+      if (extent_it != extent_map.end()) {
+        extent = extent_it->second;
+      } else {
+        extent = ref_id->extent();
+      }
+    }
+    reference_halo_extent_map[producer_or_consumer_id] = extent;
+  }
+
+  return reference_halo_extent_map;
+}
+
+//! Offset of an index of a producer axis with respect to its
+//! corresponding consumer index
+int getProducerHaloOffset(
+    const TensorView* producer_tv,
+    size_t producer_axis,
+    const TensorView* consumer_tv) {
+  auto p2c =
+      PairwiseRootDomainMap(producer_tv, consumer_tv)
+          .mapProducerToConsumer(producer_tv->domain(), consumer_tv->domain());
+
+  auto producer_id = producer_tv->getMaybeRFactorDomain()[producer_axis];
+
+  auto it = p2c.find(producer_id);
+  // Should never hit this
+  TORCH_INTERNAL_ASSERT(it != p2c.end());
```

Comment:
Hitting this would require the producer to have a reduction right?

Reply:
Yes, and that should never happen as `getProducerHaloOffset` is only used for indexing a producer and we don't use reduction axes then. I'll add more detailed comments about why it should never hit.